src/jdk.jdwp.agent/share/native/libjdwp/classTrack.c

```diff
@@ -81,15 +81,35 @@ cbTrackingObjectFree(jvmtiEnv* jvmti_env, jlong tag)
 struct bag *
 classTrack_processUnloads(JNIEnv *env)
 {
-    debugMonitorEnter(classTrackLock);
-    if (deletedSignatures == NULL) {
-        // Class tracking not initialized, nobody's interested.
-        debugMonitorExit(classTrackLock);
-        return NULL;
-    }
-    struct bag* deleted = deletedSignatures;
-    deletedSignatures = bagCreateBag(sizeof(char*), 10);
-    debugMonitorExit(classTrackLock);
+    struct bag* new_bag = NULL;
+    struct bag* deleted = NULL;
+    jboolean retry = JNI_FALSE;
+    do {
+      // Avoid unnecessary allocations when class track has yet been activated.
+      if (deletedSignatures != NULL) {
+        /* Allocate new bag outside classTrackLock lock to avoid deadlock.
+         *
+         * Note: jvmtiAllocate/jvmtiDeallocate() may be blocked by ongoing safepoints.
+         * It is dangerous to call them (via bagCreateBag/bagDestroyBag())while holding monitor(s),
+         * because jvmti may post events, e.g. JVMTI_EVENT_OBJECT_FREE at safepoints and event processing
+         * code may acquire the same monitor(s), e.g. classTrackLock in cbTrackingObjectFree(),
+         * which can lead to deadlock.
+         */
+        new_bag = bagCreateBag(sizeof(char*), 10);
+      }
+      debugMonitorEnter(classTrackLock);
+      deleted = deletedSignatures;
+      if (deletedSignatures != NULL) {
+        if (new_bag != NULL) {
+          deletedSignatures = new_bag;
+          new_bag = NULL;
+        } else {
+          retry = JNI_TRUE;
+        }
+      }
+      debugMonitorExit(classTrackLock);
+    } while (retry == JNI_TRUE);
```

Comment:
I don't think the retry is necessary. Nor is the check for `deletedSignatures != NULL`. I assume you are trying to cover the case where initially deletedSignatures was NULL (so no bag was allocated), but is not NULL by the time you grab the lock. If you return when NULL like I suggested above, you won't have this issue. Note that if not NULL, there is no way another thread can get into this same code in clear it. This is because all callers first grab the handlerLock. Grabbing of the classTrackLock here is only done to synchronize with cbTrackingObjectFree(), not with other callers of classTrack_processUnloads().

Reply:
I was worried that `classTrack_activate()` races against 'classTrack_processUnloads()` to set `deletedSignatures`, and you are right, it is also protected under ` handlerLock` lock. It simplifies a code a lot.